### PR TITLE
added support for logging in different trainer stages

### LINKED ISSUE #15794: Add support for logging in different trainer stages with DeviceStatsMonitor
### Bug description

I would like to use DeviceStatsMonitor during a trainer.test() call. I followed the [relative documentation](https://pytorch-lightning.readthedocs.io/en/stable/tuning/profiler_basic.html#measure-accelerator-usage) which makes no mention of whether this callback is exclusive to trainer.fit(). 

Despite following the docs, I get no device stats logs in my tensorboard

### How to reproduce the bug

run the following script. You will see that no stats will be logged, despite having the DeviceStatsMonitor callback

```python
import pytorch_lightning as pl
import numpy as np
from pytorch_lightning.callbacks import DeviceStatsMonitor
import torch
from torch.nn import MSELoss
from torch.optim import Adam
from torch.utils.data import DataLoader, Dataset
import torch.nn as nn


class SimpleDataset(Dataset):
    def __init__(self):
        X = np.arange(10000)
        y = X * 2
        X = [[_] for _ in X]
        y = [[_] for _ in y]
        self.X = torch.Tensor(X)
        self.y = torch.Tensor(y)

    def __len__(self):
        return len(self.y)

    def __getitem__(self, idx):
        return {"X": self.X[idx], "y": self.y[idx]}


class MyModel(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.fc = nn.Linear(1, 1)
        self.criterion = MSELoss()

    def forward(self, inputs_id, labels=None):
        outputs = self.fc(inputs_id)
        loss = 0
        if labels is not None:
            loss = self.criterion(outputs, labels)
        return loss, outputs

    def test_dataloader(self):
        dataset = SimpleDataset()
        return DataLoader(dataset, batch_size=1000)

    def test_step(self, batch, batch_idx):
        input_ids = batch["X"]
        labels = batch["y"]
        loss, outputs = self(input_ids, labels)
        return {"loss": loss}

    def configure_optimizers(self):
        optimizer = Adam(self.parameters())
        return optimizer


if __name__ == "__main__":
    model = MyModel()
    logger = pl.loggers.TensorBoardLogger(save_dir="example", name="test")
    trainer = pl.Trainer(
        logger=logger, max_epochs=5, callbacks=[DeviceStatsMonitor(cpu_stats=True)]
    )
    trainer.test(model)
```

### Environment

```
* CUDA:
	- GPU:               None
	- available:         False
	- version:           None
* Lightning:
	- pytorch-lightning: 1.7.7
	- torch:             1.12.1
	- torchmetrics:      0.10.3
* Packages:
	- absl-py:           1.3.0
	- aiohttp:           3.8.3
	- aiosignal:         1.3.1
	- antlr4-python3-runtime: 4.9.3
	- anyio:             3.6.2
	- appnope:           0.1.3
	- argon2-cffi:       21.3.0
	- argon2-cffi-bindings: 21.2.0
	- asttokens:         2.1.0
	- astunparse:        1.6.3
	- async-timeout:     4.0.2
	- attrs:             22.1.0
	- babel:             2.11.0
	- backcall:          0.2.0
	- beautifulsoup4:    4.11.1
	- bigbench:          0.0.1
	- black:             22.10.0
	- bleach:            5.0.1
	- bleurt:            0.0.2
	- cachetools:        5.2.0
	- certifi:           2022.9.24
	- cffi:              1.15.1
	- chardet:           3.0.4
	- charset-normalizer: 2.1.1
	- claficle:          0.1.0
	- click:             8.1.3
	- colorama:          0.4.6
	- contourpy:         1.0.5
	- cycler:            0.11.0
	- datasets:          2.7.0
	- debugpy:           1.6.3
	- decorator:         5.1.1
	- defusedxml:        0.7.1
	- dill:              0.3.6
	- editdistance:      0.6.0
	- entrypoints:       0.4
	- etils:             0.8.0
	- executing:         1.2.0
	- fastjsonschema:    2.16.2
	- filelock:          3.8.0
	- flake8:            4.0.1
	- flatbuffers:       2.0.7
	- fonttools:         4.37.2
	- frozenlist:        1.3.3
	- fsspec:            2022.11.0
	- future:            0.18.2
	- gast:              0.4.0
	- gin-config:        0.5.0
	- gitdb:             4.0.9
	- gitpython:         3.1.29
	- google-auth:       2.14.1
	- google-auth-oauthlib: 0.4.6
	- google-pasta:      0.2.0
	- googleapis-common-protos: 1.56.4
	- googletrans:       3.1.0a0
	- grpcio:            1.50.0
	- h11:               0.9.0
	- h2:                3.2.0
	- h5py:              3.7.0
	- hpack:             3.0.0
	- hstspreload:       2022.11.1
	- httpcore:          0.9.1
	- httpx:             0.13.3
	- huggingface-hub:   0.11.0
	- hydra-core:        1.2.0
	- hyperframe:        5.2.0
	- idna:              2.10
	- immutabledict:     2.2.1
	- importlib-metadata: 5.0.0
	- importlib-resources: 5.10.0
	- iniconfig:         1.1.1
	- ipykernel:         6.17.1
	- ipython:           8.6.0
	- ipython-genutils:  0.2.0
	- ipywidgets:        7.7.2
	- jax:               0.3.17
	- jaxlib:            0.3.15
	- jedi:              0.18.1
	- jinja2:            3.1.2
	- joblib:            1.2.0
	- json5:             0.9.10
	- jsonschema:        4.17.0
	- jupyter-client:    7.4.7
	- jupyter-core:      5.0.0
	- jupyter-server:    1.23.2
	- jupyter-server-mathjax: 0.2.6
	- jupyterlab:        3.5.0
	- jupyterlab-code-formatter: 1.5.3
	- jupyterlab-pygments: 0.2.2
	- jupyterlab-server: 2.16.3
	- jupyterlab-widgets: 1.1.1
	- keras:             2.10.0
	- keras-preprocessing: 1.1.2
	- kiwisolver:        1.4.4
	- libclang:          14.0.6
	- markdown:          3.4.1
	- markupsafe:        2.1.1
	- matplotlib:        3.6.0
	- matplotlib-inline: 0.1.6
	- mccabe:            0.6.1
	- mesh-tensorflow:   0.1.21
	- mistune:           2.0.4
	- multidict:         6.0.2
	- multiprocess:      0.70.14
	- mypy:              0.971
	- mypy-extensions:   0.4.3
	- nbclassic:         0.4.8
	- nbclient:          0.7.0
	- nbconvert:         7.2.5
	- nbdime:            3.1.1
	- nbformat:          5.7.0
	- nest-asyncio:      1.5.6
	- nltk:              3.7
	- notebook:          6.5.2
	- notebook-shim:     0.2.2
	- numpy:             1.23.4
	- oauthlib:          3.2.2
	- omegaconf:         2.2.3
	- opt-einsum:        3.3.0
	- packaging:         21.3
	- pandas:            1.5.1
	- pandocfilters:     1.5.0
	- parso:             0.8.3
	- pathspec:          0.10.2
	- pexpect:           4.8.0
	- pickleshare:       0.7.5
	- pillow:            9.2.0
	- pip:               22.1.2
	- pkgutil-resolve-name: 1.3.10
	- platformdirs:      2.5.4
	- pluggy:            1.0.0
	- portalocker:       2.5.1
	- prometheus-client: 0.15.0
	- promise:           2.3
	- prompt-toolkit:    3.0.32
	- protobuf:          3.20.3
	- psutil:            5.9.4
	- ptyprocess:        0.7.0
	- pure-eval:         0.2.2
	- py:                1.11.0
	- pyarrow:           10.0.0
	- pyasn1:            0.4.8
	- pyasn1-modules:    0.2.8
	- pycodestyle:       2.8.0
	- pycparser:         2.21
	- pydeprecate:       0.3.2
	- pyflakes:          2.4.0
	- pygments:          2.13.0
	- pyparsing:         3.0.9
	- pyrsistent:        0.19.2
	- pytest:            7.1.3
	- python-dateutil:   2.8.2
	- pytorch-lightning: 1.7.7
	- pytz:              2022.6
	- pyyaml:            6.0
	- pyzmq:             24.0.1
	- regex:             2022.10.31
	- requests:          2.28.1
	- requests-oauthlib: 1.3.1
	- requests-unixsocket: 0.3.0
	- responses:         0.18.0
	- restrictedpython:  5.2
	- rfc3986:           1.5.0
	- rouge-score:       0.1.2
	- rsa:               4.9
	- sacrebleu:         2.2.1
	- scikit-learn:      1.1.2
	- scipy:             1.9.1
	- seaborn:           0.12.0
	- send2trash:        1.8.0
	- sentencepiece:     0.1.97
	- seqio:             0.0.10
	- setuptools:        65.5.1
	- six:               1.16.0
	- smmap:             5.0.0
	- sniffio:           1.3.0
	- soupsieve:         2.3.2.post1
	- stack-data:        0.6.1
	- t5:                0.9.3
	- tabulate:          0.8.10
	- tensorboard:       2.11.0
	- tensorboard-data-server: 0.6.1
	- tensorboard-plugin-wit: 1.8.1
	- tensorflow-datasets: 4.6.0
	- tensorflow-estimator: 2.10.0
	- tensorflow-hub:    0.12.0
	- tensorflow-io-gcs-filesystem: 0.27.0
	- tensorflow-metadata: 1.10.0
	- tensorflow-text:   2.10.0
	- termcolor:         2.1.0
	- terminado:         0.17.0
	- tf-slim:           1.1.0
	- tfds-nightly:      4.6.0.dev202209160046
	- threadpoolctl:     3.1.0
	- tinycss2:          1.2.1
	- tokenizers:        0.13.2
	- toml:              0.10.2
	- tomli:             2.0.1
	- torch:             1.12.1
	- torchmetrics:      0.10.3
	- tornado:           6.2
	- tqdm:              4.64.1
	- traitlets:         5.5.0
	- transformers:      4.24.0
	- types-pyyaml:      6.0.12.2
	- typing-extensions: 4.4.0
	- urllib3:           1.26.12
	- wcwidth:           0.2.5
	- webencodings:      0.5.1
	- websocket-client:  1.4.2
	- werkzeug:          2.2.2
	- wheel:             0.38.4
	- widgetsnbextension: 3.6.1
	- wrapt:             1.14.1
	- xxhash:            3.1.0
	- yarl:              1.8.1
	- zipp:              3.10.0
* System:
	- OS:                Darwin
	- architecture:
		- 64bit
		- 
	- processor:         i386
	- python:            3.8.13
	- version:           Darwin Kernel Version 19.6.0: Tue Jun 21 21:18:39 PDT 2022; root:xnu-6153.141.66~1/RELEASE_X86_64
```

### More info

I have verified this on both GPU and CPU. The example above uses CPU.

cc @borda @awaelchli

### PR DESCRIPTION
**Pull request by yMayanand**
added support for logging in different trainer stages with DeviceStatsMonitor

## What does this PR do?

Fixes #15794

## Before submitting

- [X] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [X] Did you make sure to **update the documentation** with your changes? (if necessary)
- [X] Did you write any **new necessary tests**? (not for typos and docs)
- [X] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [X] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [X] Check that all items from **Before submitting** are resolved
- [X] Make sure the title is self-explanatory and the description concisely explains the PR
- [X] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/tests_pytorch/callbacks/test_device_stats_monitor.py [hunk 1]
```diff
@@ -166,3 +168,45 @@ def test_device_stats_monitor_warning_when_psutil_not_available(monkeypatch, tmp
...
+    trainer.test(model)
+
+    with open(f"{tmpdir}/lightning_logs/version_0/metrics.csv") as csvfile:
+
+        content = csv.reader(csvfile, delimiter=",")
+        it = iter(content).__next__()
+
+        # searching for training stage logs
+        train_stage_results = [re.match(r".+on_train_batch", i) for i in it]
+        train = any(train_stage_results)
```
Comment (carmocca) [inline tests/tests_pytorch/callbacks/test_device_stats_monitor.py:202]: Can you run the assertions after each stage instead of all together at the end? It's easier to know what failed if it does then

```suggestion
        assert any(train_stage_results)
```

File: tests/tests_pytorch/callbacks/test_device_stats_monitor.py [hunk 2]
```diff
@@ -166,3 +168,45 @@ def test_device_stats_monitor_warning_when_psutil_not_available(monkeypatch, tmp
...
+        enable_checkpointing=False,
+        enable_progress_bar=False,
+    )
+
+    trainer.fit(model)
+    trainer.test(model)
+
+    with open(f"{tmpdir}/lightning_logs/version_0/metrics.csv") as csvfile:
+
+        content = csv.reader(csvfile, delimiter=",")
```
Comment (carmocca) [inline tests/tests_pytorch/callbacks/test_device_stats_monitor.py:197]: Everything after this line can be put outside the context manager

File: tests/tests_pytorch/callbacks/test_device_stats_monitor.py [hunk 3]
```diff
@@ -166,3 +168,45 @@ def test_device_stats_monitor_warning_when_psutil_not_available(monkeypatch, tmp
...
+def test_device_stats_monitor_logs_for_different_stages(tmpdir):
+    """Test that metrics are logged for all stages that is training, testing and validation."""
+
+    model = BoringModel()
+    device_stats = DeviceStatsMonitor()
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=2,
+        limit_train_batches=7,
```
Comment (carmocca) [inline tests/tests_pytorch/callbacks/test_device_stats_monitor.py:182]: 1 epoch should be fine.
And it doesn't need 7 batches, 1 should be fine.
You should also set `limit_test_batches=1`

File: tests/tests_pytorch/callbacks/test_device_stats_monitor.py (hunk 2 above)
Comment (yMayanand) [reply tests/tests_pytorch/callbacks/test_device_stats_monitor.py:197]: Its makes sense to put unnecessary things out context manager I will update the file accordingly.

File: tests/tests_pytorch/callbacks/test_device_stats_monitor.py (hunk 3 above)
Comment (yMayanand) [reply tests/tests_pytorch/callbacks/test_device_stats_monitor.py:182]: I will change epochs and test batches for faster testing and thank you for all your suggestions 🙏.

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by carmocca — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
